File: Graph/monsters/monsters.py
# 5 8
# ########
# #M..A..#
# #.#.M#.#
# #M#..#..
# #.######

from collections import deque

if __name__ == "__main__":
    inp1 = input().split(' ')
    h, w = int(inp1[0]), int(inp1[1])
    maze = []

    for i in range(h):
        maze.append(list(input()))

    start = None
    monsters = []
    # Find monsters and start
    for i in range(h):
        for j in range(w):
            if maze[i][j] == 'M':
                monsters.append((i, j, 0))
            elif maze[i][j] == 'A':
                start = (i, j, 0)

    INF = float('inf')
    # Separate grid: earliest step a monster can reach each cell.
    monsterTime = [[INF] * w for _ in range(h)]

    dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)]

    # Monsters BFS (multi-source). Distances stay minimal because we
    # popleft (FIFO) and mark each cell's time exactly once when first seen.
    mq = deque()
    for m in monsters:
        monsterTime[m[0]][m[1]] = 0
        mq.append(m)
    while len(mq) > 0:
        e_i, e_j, step = mq.popleft()

        for d in dirs:
            ni, nj = e_i + d[0], e_j + d[1]
            if 0 <= ni < h and 0 <= nj < w:
                if maze[ni][nj] != '#' and monsterTime[ni][nj] == INF:
                    monsterTime[ni][nj] = step + 1
                    mq.append((ni, nj, step + 1))

    # A BFS. Player may enter a cell only if it arrives strictly before any
    # monster (step + 1 < monsterTime). A separate visited set dedupes cells
    # so the queue drains and parent pointers form a tree (no cycles).
    aq = deque()
    parentMap = dict()
    visited = set()
    visited.add((start[0], start[1]))
    parentMap[(start[0], start[1])] = None
    aq.append(start)

    won = False
    won_coord = None

    while len(aq) > 0:
        e_i, e_j, step = aq.popleft()

        # Reaching any boundary cell means the player has escaped.
        if e_i == 0 or e_i == h - 1 or e_j == 0 or e_j == w - 1:
            won = True
            won_coord = (e_i, e_j)
            break

        for d in dirs:
            ni, nj = e_i + d[0], e_j + d[1]
            if 0 <= ni < h and 0 <= nj < w:
                if maze[ni][nj] != '#' and (ni, nj) not in visited and step + 1 < monsterTime[ni][nj]:
                    visited.add((ni, nj))
                    parentMap[(ni, nj)] = (e_i, e_j)
                    aq.append((ni, nj, step + 1))

    if not won:
        print("NO")
    else:
        # Walk parent pointers from the boundary back to A, collecting cells.
        cells = []
        pTrack = won_coord
        while pTrack is not None:
            cells.append(pTrack)
            pTrack = parentMap[pTrack]
        cells.reverse()  # now A -> ... -> boundary

        # Convert each consecutive cell pair into a direction letter.
        moveOf = {(-1, 0): 'U', (1, 0): 'D', (0, -1): 'L', (0, 1): 'R'}
        path = []
        for k in range(1, len(cells)):
            pi, pj = cells[k - 1]
            ci, cj = cells[k]
            path.append(moveOf[(ci - pi, cj - pj)])

        print("YES")
        print(len(path))
        print("".join(path))

File: Graph/monsters/monsters.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int h, w;
    cin >> h >> w;

    vector<string> maze(h);
    for (int i = 0; i < h; i++) cin >> maze[i];

    const int INF = INT_MAX;
    vector<vector<int>> monsterTime(h, vector<int>(w, INF));

    // 4 directions with their move letters (U, R, D, L).
    const int di[4] = {-1, 0, 1, 0};
    const int dj[4] = {0, 1, 0, -1};
    const char moveCh[4] = {'U', 'R', 'D', 'L'};

    // Multi-source BFS from every monster.
    deque<pair<int,int>> mq;
    int si = -1, sj = -1;
    for (int i = 0; i < h; i++) {
        for (int j = 0; j < w; j++) {
            if (maze[i][j] == 'M') {
                monsterTime[i][j] = 0;
                mq.push_back({i, j});
            } else if (maze[i][j] == 'A') {
                si = i; sj = j;
            }
        }
    }
    while (!mq.empty()) {
        auto [ci, cj] = mq.front(); mq.pop_front();
        for (int k = 0; k < 4; k++) {
            int ni = ci + di[k], nj = cj + dj[k];
            if (ni < 0 || ni >= h || nj < 0 || nj >= w) continue;
            if (maze[ni][nj] == '#') continue;
            if (monsterTime[ni][nj] != INF) continue;
            monsterTime[ni][nj] = monsterTime[ci][cj] + 1;
            mq.push_back({ni, nj});
        }
    }

    // Player BFS from A, gated by arriving strictly before any monster.
    // parent[i][j] stores the direction index used to step INTO (i,j),
    // and visited marks cells so the queue drains and paths form a tree.
    vector<vector<int>> parentDir(h, vector<int>(w, -1));
    vector<vector<char>> visited(h, vector<char>(w, 0));
    vector<vector<int>> dist(h, vector<int>(w, 0));

    deque<pair<int,int>> aq;
    visited[si][sj] = 1;
    aq.push_back({si, sj});

    bool won = false;
    int wi = -1, wj = -1;

    while (!aq.empty()) {
        auto [ci, cj] = aq.front(); aq.pop_front();

        if (ci == 0 || ci == h - 1 || cj == 0 || cj == w - 1) {
            won = true; wi = ci; wj = cj;
            break;
        }

        for (int k = 0; k < 4; k++) {
            int ni = ci + di[k], nj = cj + dj[k];
            if (ni < 0 || ni >= h || nj < 0 || nj >= w) continue;
            if (maze[ni][nj] == '#' || visited[ni][nj]) continue;
            if (dist[ci][cj] + 1 >= monsterTime[ni][nj]) continue;  // monster gets there first
            visited[ni][nj] = 1;
            dist[ni][nj] = dist[ci][cj] + 1;
            parentDir[ni][nj] = k;
            aq.push_back({ni, nj});
        }
    }

    if (!won) {
        cout << "NO\n";
        return 0;
    }

    // Reconstruct: walk backward from the boundary cell to A using parentDir.
    string path;
    int ci = wi, cj = wj;
    while (!(ci == si && cj == sj)) {
        int k = parentDir[ci][cj];
        path.push_back(moveCh[k]);
        ci -= di[k];  // step back to the parent cell
        cj -= dj[k];
    }
    reverse(path.begin(), path.end());

    cout << "YES\n";
    cout << path.size() << "\n";
    cout << path << "\n";

    return 0;
}
